zend_abstract_interface/interceptor/php8/interceptor.c

```diff
@@ -0,0 +1,238 @@
+        zai_hook_memory_table_insert((zend_execute_data *)generator, &frame_memory);
+    }
+
+    return &generator->std;
+}
+
+static zend_object_dtor_obj_t zai_interceptor_generator_dtor_obj;
+static void zai_interceptor_generator_dtor_wrapper(zend_object *object) {
+    zend_generator *generator = (zend_generator *)object;
+
+    zai_hook_memory_t *frame_memory;
+    if (zai_hook_memory_table_find((zend_execute_data *)generator, &frame_memory)) {
+        // generator dtor frees it
+        zend_execute_data ex = *generator->execute_data;
+
+        zai_interceptor_generator_dtor_obj(object);
+
+        // may have returned in the dtor, don't execute twice
+        if (zai_hook_memory_table_find((zend_execute_data *)generator, &frame_memory)) {
+            // aborted generator
+            zval retval;
+            ZVAL_NULL(&retval);
+            zai_hook_finish(&ex, &retval, frame_memory);
+            zai_hook_memory_table_del((zend_execute_data *)generator);
+        }
+    }
+}
+
+static void (*prev_execute_internal)(zend_execute_data *execute_data, zval *return_value);
+static inline void zai_interceptor_execute_internal_impl(zend_execute_data *execute_data, zval *return_value, bool prev) {
+    HashTable *hooks;
+    zend_function *func = execute_data->func;
+    if (UNEXPECTED(zai_hook_resolved_table_find((zend_ulong)func, &hooks))) {
+        zai_hook_memory_t frame_memory;
+        if (zai_hook_continue(execute_data, &frame_memory)) {
+            zai_hook_memory_table_insert(execute_data, &frame_memory);
+        }
+
+        // we do not use try / catch here as to preserve order of hooks, LIFO style, in bailout handler
+        if (prev) {
+            prev_execute_internal(execute_data, return_value);
+        } else {
+            func->internal_function.handler(execute_data, return_value);
+        }
+
+        zai_hook_finish(execute_data, return_value, &frame_memory);
+        zai_hook_memory_table_del(execute_data);
+    } else {
+        if (prev) {
+            prev_execute_internal(execute_data, return_value);
+        } else {
+            func->internal_function.handler(execute_data, return_value);
+        }
+    }
+}
+
+static void zai_interceptor_execute_internal_no_prev(zend_execute_data *execute_data, zval *return_value) {
+    zai_interceptor_execute_internal_impl(execute_data, return_value, false);
+}
+
+static void zai_interceptor_execute_internal(zend_execute_data *execute_data, zval *return_value) {
+    zai_interceptor_execute_internal_impl(execute_data, return_value, true);
+}
+
+void zai_interceptor_minit() {
+    prev_execute_internal = zend_execute_internal;
+    zend_execute_internal = prev_execute_internal ? zai_interceptor_execute_internal : zai_interceptor_execute_internal_no_prev;
+
+    zend_observer_fcall_register(zai_interceptor_observer_fcall_init);
+
+    // get hold of a generator object to access handlers
+    zend_objects_store objects_store = EG(objects_store);
```

Comment:
super sneaky, I've never seen anything do this before and I'm not sure what side effects it might have ... it's a nice trick ...

Reply:
well, in this specific context, where we just call the generator ctor, there should not really be side effects...
And yes, sneaky, but works. :-)